api/src/main/java/io/minio/CopySource.java

```diff
@@ -0,0 +1,135 @@
+          throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, null, null, null);
+  }
+
+/**
+ * Create new CopySource for given bucket, object, offset, length and headerMap .
+ */
+  public CopySource(String bucketName, String objectName, Long offset, Long length,
+      Map<String, String> headerMap) throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, headerMap, null, null);
+  }
+
+  /**
+   * Create new CopySource for given bucket, object, offset, length, headerMap and CopyConditions.
+   */
+  public CopySource(String bucketName, String objectName, Long offset, Long length,
+      Map<String, String> headerMap, CopyConditions copyConditions) throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, headerMap, copyConditions, null);
+  }
+
+  /**
+   * CopySource new CopySource for given bucket, object, offset, length, headerMap, CopyConditions
+   * and server side encryption.
+   *
+   * @throws InvalidArgumentException
+   *          When offset or lenghth is negative
+    */
+  public CopySource(String bucketName, String objectName, Long offset, Long length, Map<String, String> headerMap,
+      CopyConditions copyConditions, ServerSideEncryption sseSource)
+  throws InvalidArgumentException {
+    if (bucketName == null) {
+      throw new InvalidArgumentException("Source bucket name cannot be empty");
+    }
+    if (objectName == null) {
+      throw new InvalidArgumentException("Source object name cannot be empty");
+    }
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    if ((offset != null && offset < 0) || (length != null && length < 0)) {
+      throw new InvalidArgumentException("Offset/Length can not be negative");
+    }
+    this.offset = offset;
+    this.length = length;
+    if (headerMap == null ) {
+      headerMap = new HashMap<String, String>();
+    }
+    headerMap.put("x-amz-copy-source", S3Escaper.encodePath(bucketName + "/"
+        + objectName));
+    this.headerMap = headerMap;
+    this.copyConditions = copyConditions;
+    this.sseSource = sseSource;
+
+  }
+
+  /**
+   * Returns bucket name.
+   */
+  public String bucketName() {
+    return bucketName;
+  }
+
+  /**
+   * Returns object name.
+   */
+  public String objectName() {
+    return objectName;
+  }
+
+  /**
+   * Returns offset.
+   */
+  public long offset() {
+    if (offset == null) {
+      return 0L;
+    }
+    return offset;
+  }
+
+  /**
+   * Returns length.
+   */
+  public long length() {
+    if (length == null) {
+      length = 0L;
+    }
+    return length;
+  }
+  /**
+   * Returns headerMap.
+   */
+
+  public Map<String, String> headers() {
+    if ( headerMap ==   null ) {
+      headerMap = new HashMap<String,String>();
+    }
+    // If no conditions available, skip addition else add the conditions to the header
+    if (copyConditions != null) {
+      headerMap.putAll(copyConditions.getConditions());
```

Comment:
You are adding to `headerMap` every time this method is called. Similarly initialization happens too. This method should just return `headerMap` and updating `headerMap` should be done in `public CopySource(String bucketName, String objectName, Long offset, Long length, Map<String, String> headerMap,
      CopyConditions copyConditions, ServerSideEncryption sseSource)`

Reply:
done.